File: android/src/main/java/com/margelo/nitro/nitrolocationtracking/LocationQueue.kt
package com.margelo.nitro.nitrolocationtracking

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log

/**
 * Durable FIFO queue of pending Live Push bodies, backed by SQLite so it
 * survives app kill / reboot. Each row is one serialized POST body (the
 * per-fix JSON object). The drainer in [LivePusher] peeks the oldest rows,
 * POSTs them, and deletes only on success.
 *
 * All methods are synchronous and thread-safe (SQLiteDatabase serializes its
 * own access); callers invoke them from LivePusher's single drain executor.
 */
class LocationQueue(context: Context) {

  companion object {
    private const val TAG = "NitroLocationQueue"
    private const val DB_NAME = "nitro_livepush_queue.db"
    private const val DB_VERSION = 1
    private const val TABLE = "pending"
    private const val COL_ID = "id"
    private const val COL_BODY = "body"
    private const val COL_CREATED = "created_at"
  }

  private class Helper(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {
    override fun onCreate(db: SQLiteDatabase) {
      db.execSQL(
        "CREATE TABLE $TABLE ($COL_ID INTEGER PRIMARY KEY AUTOINCREMENT, " +
          "$COL_BODY TEXT NOT NULL, $COL_CREATED INTEGER NOT NULL)"
      )
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
      // v1 only — nothing to migrate yet.
    }
  }

  /** One pending row: its primary key and serialized body. */
  data class Row(val id: Long, val body: String)

  private val helper = Helper(context.applicationContext)

  fun enqueue(body: String, nowMs: Long) {
    try {
      val values = ContentValues().apply {
        put(COL_BODY, body)
        put(COL_CREATED, nowMs)
      }
      helper.writableDatabase.insert(TABLE, null, values)
    } catch (e: Exception) {
      Log.w(TAG, "enqueue failed: ${e.message}")
    }
  }

  /** Oldest [limit] rows, FIFO order. */
  fun peek(limit: Int): List<Row> {
    if (limit <= 0) return emptyList()
    val rows = ArrayList<Row>(limit)
    try {
      helper.readableDatabase.query(
        TABLE, arrayOf(COL_ID, COL_BODY), null, null, null, null,
        "$COL_ID ASC", limit.toString()
      ).use { c ->
        val idIdx = c.getColumnIndexOrThrow(COL_ID)
        val bodyIdx = c.getColumnIndexOrThrow(COL_BODY)
        while (c.moveToNext()) rows.add(Row(c.getLong(idIdx), c.getString(bodyIdx)))
      }
    } catch (e: Exception) {
      Log.w(TAG, "peek failed: ${e.message}")
    }
    return rows
  }

  fun delete(ids: List<Long>) {
    if (ids.isEmpty()) return
    try {
      val placeholders = ids.joinToString(",") { "?" }
      val args = ids.map { it.toString() }.toTypedArray()
      helper.writableDatabase.delete(TABLE, "$COL_ID IN ($placeholders)", args)
    } catch (e: Exception) {
      Log.w(TAG, "delete failed: ${e.message}")
    }
  }

  fun count(): Int {
    return try {
      helper.readableDatabase.rawQuery("SELECT COUNT(*) FROM $TABLE", null).use { c ->
        if (c.moveToFirst()) c.getInt(0) else 0
      }
    } catch (e: Exception) {
      Log.w(TAG, "count failed: ${e.message}")
      0
    }
  }

  /** Drop oldest rows so the queue never exceeds [maxSize]. */
  fun trimToMax(maxSize: Int) {
    if (maxSize <= 0) return
    try {
      val excess = count() - maxSize
      if (excess <= 0) return
      // Delete the `excess` oldest ids.
      helper.writableDatabase.execSQL(
        "DELETE FROM $TABLE WHERE $COL_ID IN " +
          "(SELECT $COL_ID FROM $TABLE ORDER BY $COL_ID ASC LIMIT $excess)"
      )
      Log.d(TAG, "trimmed $excess oldest rows (cap=$maxSize)")
    } catch (e: Exception) {
      Log.w(TAG, "trimToMax failed: ${e.message}")
    }
  }
}

File: ios/LocationQueue.swift
import Foundation
import SQLite3

/**
 * Durable FIFO queue of pending Live Push bodies, backed by SQLite (the system
 * libsqlite3) so it survives app kill / reboot. Each row is one serialized POST
 * body. The drainer in `LivePusher` peeks the oldest rows, POSTs them, and
 * deletes only on success.
 *
 * Not internally synchronized — `LivePusher` calls every method from its single
 * serial drain queue, so access is already serialized.
 */
final class LocationQueue {

  struct Row { let id: Int64; let body: String }

  // sqlite3 wants SQLITE_TRANSIENT so it copies bound text; the symbol isn't
  // imported into Swift, so reconstruct it.
  private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

  private var db: OpaquePointer?

  init() {
    let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
      ?? FileManager.default.temporaryDirectory
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let path = dir.appendingPathComponent("nitro_livepush_queue.sqlite").path

    if sqlite3_open(path, &db) == SQLITE_OK {
      exec(
        "CREATE TABLE IF NOT EXISTS pending (" +
        "id INTEGER PRIMARY KEY AUTOINCREMENT, body TEXT NOT NULL, created_at INTEGER NOT NULL)"
      )
    } else {
      NSLog("[NitroLocationQueue] failed to open db at \(path)")
      db = nil
    }
  }

  deinit {
    if db != nil { sqlite3_close(db) }
  }

  func enqueue(body: String, nowMs: Int64) {
    guard let db else { return }
    var stmt: OpaquePointer?
    defer { sqlite3_finalize(stmt) }
    guard sqlite3_prepare_v2(db, "INSERT INTO pending (body, created_at) VALUES (?, ?)", -1, &stmt, nil) == SQLITE_OK else { return }
    sqlite3_bind_text(stmt, 1, body, -1, SQLITE_TRANSIENT)
    sqlite3_bind_int64(stmt, 2, nowMs)
    if sqlite3_step(stmt) != SQLITE_DONE {
      NSLog("[NitroLocationQueue] enqueue failed")
    }
  }

  /// Oldest `limit` rows, FIFO order.
  func peek(limit: Int) -> [Row] {
    guard let db, limit > 0 else { return [] }
    var stmt: OpaquePointer?
    defer { sqlite3_finalize(stmt) }
    var rows: [Row] = []
    guard sqlite3_prepare_v2(db, "SELECT id, body FROM pending ORDER BY id ASC LIMIT ?", -1, &stmt, nil) == SQLITE_OK else { return [] }
    sqlite3_bind_int(stmt, 1, Int32(limit))
    while sqlite3_step(stmt) == SQLITE_ROW {
      let id = sqlite3_column_int64(stmt, 0)
      let body = sqlite3_column_text(stmt, 1).map { String(cString: $0) } ?? ""
      rows.append(Row(id: id, body: body))
    }
    return rows
  }

  func delete(ids: [Int64]) {
    guard let db, !ids.isEmpty else { return }
    let placeholders = ids.map { _ in "?" }.joined(separator: ",")
    var stmt: OpaquePointer?
    defer { sqlite3_finalize(stmt) }
    guard sqlite3_prepare_v2(db, "DELETE FROM pending WHERE id IN (\(placeholders))", -1, &stmt, nil) == SQLITE_OK else { return }
    for (i, id) in ids.enumerated() {
      sqlite3_bind_int64(stmt, Int32(i + 1), id)
    }
    if sqlite3_step(stmt) != SQLITE_DONE {
      NSLog("[NitroLocationQueue] delete failed")
    }
  }

  func count() -> Int {
    guard let db else { return 0 }
    var stmt: OpaquePointer?
    defer { sqlite3_finalize(stmt) }
    guard sqlite3_prepare_v2(db, "SELECT COUNT(*) FROM pending", -1, &stmt, nil) == SQLITE_OK else { return 0 }
    return sqlite3_step(stmt) == SQLITE_ROW ? Int(sqlite3_column_int(stmt, 0)) : 0
  }

  /// Drop oldest rows so the queue never exceeds `maxSize`.
  func trimToMax(_ maxSize: Int) {
    guard maxSize > 0 else { return }
    let excess = count() - maxSize
    guard excess > 0 else { return }
    exec("DELETE FROM pending WHERE id IN (SELECT id FROM pending ORDER BY id ASC LIMIT \(excess))")
  }

  private func exec(_ sql: String) {
    guard let db else { return }
    if sqlite3_exec(db, sql, nil, nil, nil) != SQLITE_OK {
      NSLog("[NitroLocationQueue] exec failed: \(sql)")
    }
  }
}
